src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/LSPRazorProjectHost.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [AppliesTo("(DotNetCoreRazor & DotNetCoreRazorConfiguration) | ((DotNetCoreRazor | DotNetCoreWeb) & !DotNetCoreRazorConfiguration)")]
+    [Export(ExportContractNames.Scopes.UnconfiguredProject, typeof(IProjectDynamicLoadComponent))]
+    internal class LSPRazorProjectHost : IProjectDynamicLoadComponent
+    {
+        private readonly LSPEditorFeatureDetector _featureDetector;
+        private readonly List<Lazy<ILanguageClient, ILanguageClientMetadata>> _applicableLanguageClients;
+        private readonly Lazy<ILanguageClientBroker> _languageClientBroker;
+
+        [ImportingConstructor]
+        public LSPRazorProjectHost(
+            LSPEditorFeatureDetector featureDetector,
+            Lazy<ILanguageClientBroker> languageClientBroker,
+            [ImportMany] IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            if (featureDetector is null)
+            {
+                throw new ArgumentNullException(nameof(featureDetector));
+            }
+
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (languageClients is null)
+            {
+                throw new ArgumentNullException(nameof(languageClients));
+            }
+
+            _featureDetector = featureDetector;
+            _languageClientBroker = languageClientBroker;
+            _applicableLanguageClients = GetApplicableClients(languageClients);
+        }
+
+        public Task LoadAsync()
+        {
+            if (!_featureDetector.IsLSPEditorFeatureEnabled())
+            {
+                return Task.CompletedTask;
+            }
+
+            foreach (var client in _applicableLanguageClients)
+            {
+                _languageClientBroker.Value.LoadAsync(client.Metadata, client.Value).Forget();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task UnloadAsync()
+        {
+            return Task.CompletedTask;
```

Comment:
Just for my knowledge; why don't we call for the language client broker / applicable language servers to be individually unloaded?

Reply:
They don't provide a way for us to do that. I believe this is because they want to control the lifetime of the servers. They will invoke a shutdown request if it needs to be unloaded.